### PR TITLE
Optimize async invocation

### PR DESCRIPTION
**Pull request by geoffkizer**
Move ExecutionContext and SynchronizationContext to Thread object to optimize their access, particularly for ExecutionContextSwitcher.

Tweak some ExecutionContextSwitcher related code to optimize it a bit further.

### CODE REVIEW

**Comment by geoffkizer** [comment]
@jkotas, @ericeil

File: src/mscorlib/src/System/Threading/SynchronizationContext.cs [hunk 1]
```diff
@@ -244,7 +241,7 @@ private static SynchronizationContext GetThreadLocalContext()
             SynchronizationContext context = null;
             
 #if FEATURE_APPX
-            if (context == null && Environment.IsWinRTSupported)
+            if (context == null && Environment.IsWinRTSupported && AppDomain.IsAppXModel())
```
Comment (jkotas) [inline src/mscorlib/src/System/Threading/SynchronizationContext.cs]: You can remove the `Environment.IsWinRTSupported` check. `AppDomain.IsAppXModel` is sufficient.

**Comment by jkotas** [comment]
LGTM otherwise. Could you please sign the cla to make everything green?

**Comment by jkotas** [comment]
Also, it may be useful to run corefx tests prior to checkin if you have not done it already - https://github.com/dotnet/coreclr/blob/master/Documentation/building/testing-with-corefx.md. The corefx tests have better coverage for the framework functionality like this one.

**Comment by omariom** [comment]
Does it make async invocation faster?

**Comment by ericeil** [comment]
cc:@kouvel.  

LGTM.

**Comment by kouvel** [comment]
LGTM as well

**Comment by geoffkizer** [comment]
Perf improvement:

Microbenchmark code:

    public static Task g_task;

    public static async Task AsyncMethod()
    {
        await g_task;
    }

    public static async Task DoAsyncTest()
    {
        for (int i = 0; i < 100000000; i++)
            await AsyncMethod();
    }

    public static void Test()
    {
        Console.WriteLine("Running Test");

        TaskCompletionSource<int> tcs = new TaskCompletionSource<int>();
        tcs.SetResult(0);

        g_task = tcs.Task;

        DateTime start = DateTime.Now;
        
        DoAsyncTest().Wait();

        DateTime stop = DateTime.Now;
        Console.WriteLine("Elapsed = {0}", stop - start);
    }

This improved from 0.90s to 0.45s.

Also, TechEmpower Plaintext throughput improves by about 10%.

**Pull request merged**